### PR TITLE
Fix rate.match is not a function issue

### LINKED ISSUE #20: TypeError: schedule.rate.match is not a function
Hi, this error started today on my nightly deploy task:

```
Serverless: Converting local crontabs to UTC crontabs...
 
 Type Error ----------------------------------------------
 
  TypeError: schedule.rate.match is not a function
      at ServerlessLocalCrontabs.convertCrontabs (/root/project/build/node_modules/serverless-local-schedule/index.js:73:37)
      at PluginManager.invoke (/root/project/build/node_modules/serverless/lib/classes/PluginManager.js:580:20)
      at async PluginManager.spawn (/root/project/build/node_modules/serverless/lib/classes/PluginManager.js:602:5)
      at async Object.before:deploy:deploy [as hook] (/root/project/build/node_modules/serverless/lib/plugins/deploy.js:61:11)
      at async PluginManager.invoke (/root/project/build/node_modules/serverless/lib/classes/PluginManager.js:580:9)
      at async PluginManager.run (/root/project/build/node_modules/serverless/lib/classes/PluginManager.js:640:7)
      at async Serverless.run (/root/project/build/node_modules/serverless/lib/Serverless.js:457:5)
      at async /root/project/build/node_modules/serverless/scripts/serverless.js:803:9
```

I believe it relates to this bit of configuration in serverless.yml:

```
  subscriptions-f:
    environment: ${file(.env.json)}
    handler: dist/subscriptions-f-handler.handler
    timeout: 900
    events:
      - sns:
          arn: ${file(.env.json):USER_CHAIN_TOPIC}
      - schedule:
          rate: cron(0 * * * ? *)
          timezone: Europe/London
          input:
            topic: 'main'
```

I was able to deploy the same code to UAT and to Production 2 days ago - this issue started on nightly provisioning only.

I don't know if it matters, I have been using serverless-local-schedule for about 1 year but with frameworkVersion < 2.
Now I'm using frameworkVersion: "2"

Any idea? Thank you!

### PR DESCRIPTION
**Pull request by ejoncas-rivalbet**
FIXES: https://github.com/UnitedIncome/serverless-local-schedule/issues/20

Solution by https://github.com/UnitedIncome/serverless-local-schedule/issues/20#issuecomment-924749622

### CODE REVIEW

**Comment by kotowick** [comment]
Any update to when this will be merged?

**Comment by steresi** [comment]
I'm also eager to have this fix merged, to work with newer Serverless!

File: index.js [hunk 1]
```diff
@@ -70,7 +70,8 @@ function convertCrontabs() {
         event.schedule.hasOwnProperty("timezone")
       ) {
         const schedule = event.schedule;
-        const match = schedule.rate.match(/^cron\((.*)\)$/);
+        const rate = Array.isArray(schedule.rate) ? schedule.rate[0] : schedule.rate;
```
Comment (pgrzesik) [inline index.js:73]: I'm not familiar with this codebase but I'm afraid this approach might have a bug. If the rate is an array, I believe we should convert all entries in an array e.g. for config like this:

```
functions:
  foo:
    handler: foo.handler
    events:
      - schedule:
          rate:
            - cron(0 0/4 ? * MON-FRI *)
            - cron(0 2 ? * SAT-SUN *)
```

The proposed approach would only address the first `cron`.

**Review by pgrzesik — CHANGES_REQUESTED** [review]
I believe there's a bug in the proposed approach, please let me know what do you think

File: index.js (hunk 1 above)
Comment (schadilibre) [reply index.js:73]: You're right, the fix works fine with one schedule but breaks with multiple ones.

I wish I had time to propose a solution for the case with multiple schedules, but for now it's complicated... 😬

**Comment by akshitkh** [comment]
Any update on this? We are unable to use this with the latest Serverless version.

File: index.js (hunk 1 above)
Comment (ejoncas-rivalbet) [reply index.js:73]: multiples cron in the same rate is not a valid servelress configuration. This is just a plugin that resolves the timezone component, but the main configuration needs to follow the serverless configuration: 

See for more details https://www.serverless.com/framework/docs/providers/aws/events/schedule

Multiple rate entries would work fine though:

```
	
functions:
  crawl:
    handler: crawl
    events:
      - schedule: cron(0 15 * * ? *)
      - schedule: cron(0 12 * * ? *)
```


I understand this is not obvious to someone reading the proposed change, but it's important to understand this plugin is not working with latest versions of serverless and this change gives you a fix for it.

I would personally go ahead as-is and if required create another ticket to support multiple cron expressions (not needed IMO).

Comment (pgrzesik) [reply index.js:73]: Hey @ejoncas-rivalbet - sorry for not responding sooner, I've missed your comment. 

> multiples cron in the same rate is not a valid servelress configuration.

Why is that the case? I was testing it and it's working properly with syntax like this:

```
events:
      - schedule:
          rate:
            - cron(0 0/4 ? * MON-FRI *)
            - cron(0 2 ? * SAT-SUN *)
```

are you observing a different behavior? It's even listed as a valid approach in the docs that you've linked to: https://www.serverless.com/framework/docs/providers/aws/events/schedule#specify-multiple-schedule-expressions

As for going forward - given the fact that supporting multiple cron expressions is a small change in code, I think it should be done as a part of this PR.

**Comment by KillDozerX2** [comment]
I'm taking over to finishing this and will also solve #13 as well

**Comment by KillDozerX2** [comment]
New PR created for @types/serverless, please upvote so it can move along as well.
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/59637

**Comment by ramblingenzyme** [comment]
https://github.com/serverless/serverless-local-schedule/pull/28 Is the new new hotness on the block to fix this issue.

**Comment by pgrzesik** [comment]
Fixed by #28

**Pull request closed without merging**